io-engine/src/grpc/v1/stats.rs

```diff
@@ -0,0 +1,162 @@
+use ::function_name::named;
+use std::panic::AssertUnwindSafe;
+
+/// RPC service for Resource IoStats.
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct StatsService {
+    name: String,
+    client_context: tokio::sync::Mutex<Option<GrpcClientContext>>,
+}
+
+#[async_trait::async_trait]
+impl<F, T> Serializer<F, T> for StatsService
+where
+    T: Send + 'static,
+    F: core::future::Future<Output = Result<T, Status>> + Send + 'static,
+{
+    async fn locked(&self, ctx: GrpcClientContext, f: F) -> Result<T, Status> {
+        let mut context_guard = self.client_context.lock().await;
+
+        // Store context as a marker of to detect abnormal termination of the
+        // request. Even though AssertUnwindSafe() allows us to
+        // intercept asserts in underlying method strategies, such a
+        // situation can still happen when the high-level future that
+        // represents gRPC call at the highest level (i.e. the one created
+        // by gRPC server) gets cancelled (due to timeout or somehow else).
+        // This can't be properly intercepted by 'locked' function itself in the
+        // first place, so the state needs to be cleaned up properly
+        // upon subsequent gRPC calls.
+        if let Some(c) = context_guard.replace(ctx) {
+            warn!("{}: gRPC method timed out, args: {}", c.id, c.args);
+        }
+
+        let fut = AssertUnwindSafe(f).catch_unwind();
+        let r = fut.await;
+
+        // Request completed, remove the marker.
+        let ctx = context_guard.take().expect("gRPC context disappeared");
+
+        match r {
+            Ok(r) => r,
+            Err(_e) => {
+                warn!("{}: gRPC method panicked, args: {}", ctx.id, ctx.args);
+                Err(Status::cancelled(format!(
+                    "{}: gRPC method panicked",
+                    ctx.id
+                )))
+            }
+        }
+    }
+}
+
+impl Default for StatsService {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl StatsService {
+    /// Constructor for Stats Service.
+    pub fn new() -> Self {
+        Self {
+            name: String::from("StatsSvc"),
+            client_context: tokio::sync::Mutex::new(None),
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl StatsRpc for StatsService {
+    #[named]
+    async fn get_pool_io_stats(
+        &self,
+        request: Request<ListStatsOption>,
+    ) -> GrpcResult<PoolIoStatsResponse> {
+        self.locked(
+            GrpcClientContext::new(&request, function_name!()),
+            async move {
+                let args = request.into_inner();
+                let rx = rpc_submit::<_, _, CoreError>(async move {
+                    let mut pool_stats: Vec<IoStats> = Vec::new();
+                    if let Some(name) = args.name {
+                        if let Some(l) = Lvs::lookup(&name) {
+                            let lvs_bdev = l.base_bdev();
+                            let stats = lvs_bdev.stats_async().await?;
```

Comment:
let lvs_bdev = l.base_bdev();
let stats = lvs_bdev.stats_async().await?;
```suggestion
let stats = l.base_bdev().lvs_bdev.stats_async().await?;                            
```